libmate-desktop/mate-colorsel.c

```diff
@@ -1371,9 +1315,14 @@ do_popup (MateColorSelection *colorsel,
 
   gtk_widget_show_all (mi);
 
-  gtk_menu_popup (GTK_MENU (menu), NULL, NULL,
-                  popup_position_func, drawing_area,
-                  3, timestamp);
+  if (trigger_event && gdk_event_triggers_context_menu (trigger_event))
+    gtk_menu_popup_at_pointer (GTK_MENU (menu), trigger_event);
+  else
+    gtk_menu_popup_at_widget (GTK_MENU (menu),
+                              drawing_area,
+                              GDK_GRAVITY_CENTER,
+                              GDK_GRAVITY_NORTH_WEST,
+                              trigger_event);
```

Comment:
Unfortunately there seem to be a GTK bug here, if you use the keyboard to popup it appears off the window…
Using `gtk_widget_get_parent (drawing_area)` as the widget works, but it **is not the right solution**.

I'll try and see to report this upstream if I can understand something around it, unless anybody knows what's going on here?

Reply:
If you replace ```gtk_drawing_area_new()``` with ```gtk_button_new```, everything will be fine. I don't know if this is a bug in GTK, it will take time to verify

Reply:
I have written test code, and this bug seems to only appear in ```gtk_drawing_area_new```
I plan to rollback the code

Reply:
Using ```gtk_widget_get_parent (drawing_area) ``` is the correct way to handle the issue. I checked the ```GTK``` code and found the reason for the incorrect menu position，```gtk_menu_popup_at_widget``` is suitable for widgets that do not create windows for themselves.
For example, ```flowbox```, ```drawarea```, and ```eventbox``` use ```gtk_widget_set_has_window (widget, TRUE)```, which can cause the menu to pop up in the wrong position

Reply:
The parent of ```drawarea``` is a ```frame```, so there is no problem using the parent
```
  frame = gtk_frame_new (NULL);
  custom_palette= palette_new (colorsel);
  gtk_container_add (GTK_CONTAINER (frame), custom_palette);
  gtk_grid_attach (GTK_GRID (grid), frame, i, j, 1, 1);
```

Reply:
> Using `gtk_widget_get_parent (drawing_area) ` is the correct way to handle the issue.

After research, in our case, it's probably the best solution indeed. It however only works because the parent is a frame that only contains the drawing area, so placing in the center of the frame is also the center of the drawing area; and even if we placed at the bottom-left it'd be only off by the borders of the frame (which might even be prettier).

But this is not a generic fix that would work in other situations though.  I just submitted the [issue upstream](https://gitlab.gnome.org/GNOME/gtk/-/issues/6166).
A usually better solution would be not having our own window on the widget, but GTK requires one to be a drag source, so we actually need one.

> I checked the `GTK` code and found the reason for the incorrect menu position，`gtk_menu_popup_at_widget` is suitable for widgets that do not create windows for themselves.

Yeah, it is basically broken if the widget's window has a non-zero position.